File: lcof-python/Tree/16_树的子结构.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
#
# 代码中的类名、方法名、参数名已经指定，请勿修改，直接返回方法规定的值即可
#
# 
# @param pRoot1 TreeNode类 
# @param pRoot2 TreeNode类 
# @return bool布尔型
#
class Solution:
    def isSame(self, pRoot1, pRoot2):
         if not pRoot2: return True
         if not pRoot1 or pRoot1.val != pRoot2.val: return False
         return self.isSame(pRoot1.left, pRoot2.left) and self.isSame(pRoot1.right, pRoot2.right)
        
    def HasSubtree(self , pRoot1: TreeNode, pRoot2: TreeNode) -> bool:
        # write code here
        if not pRoot2:
            return False
        if pRoot1 is None and pRoot2 is not None:
            return False
        if pRoot2 is None and pRoot1 is None:
            return True
        return self.isSame(pRoot1, pRoot2) or self.HasSubtree(pRoot1.left, pRoot2) or self.HasSubtree(pRoot1.right,pRoot2)

File: lcof-cpp/offer/2_Tree/16_树的子结构.cpp
/*
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) :
            val(x), left(NULL), right(NULL) {
    }
};*/
class Solution
{
public:
    bool isSame(TreeNode *root, TreeNode *node)
    {
        // 节点没了肯定是true
        if (!node)
            return true;
        // 前面已经判定node存在 root没了 返回false
        if (!root)
            return false;
        if (root->val != node->val)
            return false;

        return isSame(root->left, node->left) && isSame(root->right, node->right);
    }
    bool HasSubtree(TreeNode *pRoot1, TreeNode *pRoot2)
    {
        if (!pRoot1)
            return false;
        // pRoot2是没有的 但是上一个if 得到的pRoot1是存在的
        if (!pRoot2)
            return false;
        return isSame(pRoot1, pRoot2) || HasSubtree(pRoot1->left, pRoot2) || HasSubtree(pRoot1->right, pRoot2);
    }
};
